keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +843,117 @@ examples:
+In Alpha sidecar containers will be terminated as regular containers. No special
 or additional signals will be supported.
 
-We will collect feedback from alpha implementation and plan to improve
-termination in Beta. When Pods with sidecars are terminated:
+In Beta we have thought about introducing additional termination grace period fields
+to manage termination duration
+([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
+and leverage these fields to add reverse order termination of sidecar containers
+after the primary containers terminate.
+
+However, we decided on an alternative that doesn't require additional fields or hooks while keeping
+the desired behaviors when Pods with sidecars are terminated. While original approach works better
+with truly graceful termination where consistency is more important than time taken, proposed approach
+works for that scenario as well as a more and more popular scenario of limited time to terminate when
+graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
+   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
+   the Pod's termination grace period countdown.
+2. Once the last primary container terminates, the last started sidecar container is notified by
+   sending a `SIGTERM` signal.
+3. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous
+   sidecar container terminates.
+4. This continues until all sidecar containers have terminated, or the Pod's termination grace period
+   expires.
+5. In the latter case, all remaining containers are notified by a `SIGTERM`, followed by a fixed
+   grace period of 2s and finally terminated.
+6. The Pod will be terminated after that.
+
+Pseudocode for the above:
+
+```
+func terminatePod() {
+  // notify all sidecar containers with preStop hook, asynchronously
+  for sidecar in sidecarContainers {
+    if sidecar has preStop hook {
+      go execute preStop hook // async
+    }
+  }
+  // notify all containers with preStop hook and then SIGTERM, asynchronously
+  for container in containers {
+    if container has preStop hook {
+      go func(container) { // async
+        execute preStop hook
+        send SIGTERM
+      }
+    }
+  }
+  for {
+    switch {
+      case grace period expired:
+        for anyContainer in sidecarContainers + containers {
+          if anyContainer is running {
+            send SIGTERM
+          }
+        }
+        sleep 2s
+        for anyContainer in sidecarContainer + containers {
+          if anyContainer is running {
+            send SIGKILL
+          }
+        }
```

Comment:
I think we can simplify the code by introducing the `stopContainer`.

```go
// stopContainer is a remote call to the container runtime
// to stop the container.
function stopContainer(c container, tgps int) {
  // container runtime sends SIGTERM to the container
  // and waits for it to stop within the tgps.
  // If not killed, the container runtime will send SIGKILL to kill the container.
}
```

Reply:
maybe... does it still make it clear what's happening to all containers in case of TGPS expiration? because this was one of the main goals - I have the feeling it hides it by adding the call to function

Reply:
Separating the one remote call (it is actually one remote call and kubernetes cannot control its behavior) into multiple commands makes it a bit complicated and inaccurate.
- regular containers should not get SIGTERM twice at L900.
- regular containers should not wait for extra 2s if preStop does not consume all TGPS at L903.

nit: If you think the detail is not that important, I am okay with that.

Reply:
fair enough for the 2s delay before SIGKILL

Reply:
> please realize that there is a `if anyContainer is running {` before sending SIGTERM

Got it, you mean running = not terminating nor terminated.

Reply:
IMO this psuedocode added SO MUCH clarity.  Thank you